samples/csharp/getting-started/Ranking_AutoML/Ranking/Ranking/Program.cs

```diff
@@ -0,0 +1,175 @@
+        private static RankingData sampleRankingData = null;
+
+        static void Main(string[] args)
+        {
+            var mlContext = new MLContext(seed: 0);
+
+            // Create, train, evaluate and save a model
+            BuildTrainEvaluateAndSaveModel(mlContext);
+
+            // Make a single test prediction loading the model from .ZIP file
+            TestSinglePrediction(mlContext);
+
+            Console.WriteLine("=============== End of process, hit any key to finish ===============");
+            Console.ReadKey();
+        }
+
+        private static ITransformer BuildTrainEvaluateAndSaveModel(MLContext mlContext)
+        {
+            // STEP 1: Download and load the data
+            GetData(InputPath, OutputPath, TrainDatasetPath, TrainDatasetUrl, TestDatasetUrl, TestDatasetPath,
+                ValidationDatasetUrl, ValidationDatasetPath);
+
+            IDataView trainDataView = mlContext.Data.LoadFromTextFile<RankingData>(TrainDatasetPath, hasHeader: true, separatorChar: '\t');
+            IDataView testDataView = mlContext.Data.LoadFromTextFile<RankingData>(TestDatasetPath, hasHeader: true, separatorChar: '\t');
+
+            // STEP 2: Display first few rows of training data
+            ConsoleHelper.ShowDataViewInConsole(mlContext, trainDataView);
+
+            // STEP 3: Initialize our user-defined progress handler that AutoML will 
+            // invoke after each model it produces and evaluates.
+            var progressHandler = new RankingExperimentProgressHandler();
+
+            // STEP 4: Run AutoML ranking experiment
+            ConsoleHelper.ConsoleWriteHeader("=============== Running AutoML experiment ===============");
+            Console.WriteLine($"Running AutoML ranking experiment for {ExperimentTime} seconds...");
+
+            ExperimentResult<RankingMetrics> experimentResult = mlContext.Auto()
+                .CreateRankingExperiment(ExperimentTime)
+                .Execute(trainDataView, progressHandler: progressHandler);
+
+            // Print top models found by AutoML
+            Console.WriteLine();
+            PrintTopModels(experimentResult);
+
+            // STEP 5: Evaluate the model and print metrics
+            ConsoleHelper.ConsoleWriteHeader("=============== Evaluating model's accuracy with test data ===============");
+            RunDetail<RankingMetrics> bestRun = experimentResult.BestRun;
+
+            ITransformer trainedModel = bestRun.Model;
+            var predictions = trainedModel.Transform(testDataView);
+
+            // Get a row for single prediction
+            sampleRankingData = mlContext.Data.CreateEnumerable<RankingData>(mlContext.Data.TakeRows(testDataView, 1), reuseRowObject: false)
+                .FirstOrDefault();
+
+            var metrics = mlContext.Ranking.Evaluate(predictions);
+
+            ConsoleHelper.PrintRankingMetrics(bestRun.TrainerName, metrics);
+
+            // STEP 6: Save/persist the trained model to a .ZIP file
```

Comment:
If you're up for it, you could demonstrate thrice training. 

The general idea is to progressively refit the best found pipeline with merged data from all of the datasets in a three step process.

step | fitting data | scoring data | notes
-----|------|---|----
1 | train | valid | Done within AutoML to find the best pipeline
2 | train+valid | test | Gives final metrics which estimate the model's perf in production
3 | train+valid+test | N/A | Gives final model to launch to production

I have an example created for another [bug](https://github.com/dotnet/machinelearning/issues/5320#issuecomment-662940586): 
full example -- https://dotnetfiddle.net/nWpCkP
```c#
// Re-fit best pipeline on train and validation data, to produce 
// a model that is trained on as much data as is available while
// still having test data for the final estimate of how well the
// model will do in production.
Console.WriteLine("\n===== Refitting on train+valid and evaluating model's rsquared with test data =====");
var TrainPlusValidationDataView = textLoader.Load(new MultiFileSource(TrainDataPath, ValidationDataPath));
var refitModel1 = experimentResult.BestRun.Estimator.Fit(TrainPlusValidationDataView);
IDataView predictionsRefitOnTrainPlusValidation = refitModel1.Transform(TestDataView);
var metricsRefitOnTrainPlusValidation = mlContext.Regression.Evaluate(predictionsRefitOnTrainPlusValidation, labelColumnName: "Label", scoreColumnName: "Score");
Console.WriteLine("|" + $"{"-",-4} {experimentResult.BestRun.TrainerName,-35} {metricsRefitOnTrainPlusValidation?.RSquared ?? double.NaN,8:F4} {metricsRefitOnTrainPlusValidation?.MeanAbsoluteError ?? double.NaN,13:F2} {metricsRefitOnTrainPlusValidation?.MeanSquaredError ?? double.NaN,12:F2} {metricsRefitOnTrainPlusValidation?.RootMeanSquaredError ?? double.NaN,8:F2} {"-",9}".PadRight(112) + "|");

// Re-fit best pipeline on train, validation, and test data, to 
// produce a model that is trained on as much data as is available.
// This is the final model that can be deployed to production.
Console.WriteLine("\n===== Refitting on train+valid+test to get the final model to launch to production =====");
var TrainPlusValidationPlusTestDataView = textLoader.Load(new MultiFileSource(TrainDataPath, ValidationDataPath, TestDataPath));
var refitModel2 = experimentResult.BestRun.Estimator.Fit(TrainPlusValidationPlusTestDataView);
```

Reply:
This is an interesting concept! I'll definitely give this a go.

Reply:
Thanks for letting me know about this! I think a video on this would be good to make. 😄

Reply:
This would make a great video. This process is most beneficial when the training set is small, or the dataset is split by time. 

Ranking datasets are often split by time with the older data in the training split, newer data in the validation, and newest in the test dataset split. There's mainly two gains in using time splits; (1) removing leakage -- see "time leakage" in https://en.wikipedia.org/wiki/Leakage_(machine_learning), and (2) the most valuable data is the most up to date data as user trends shift over time.

The newest data is the most representative of the live production traffic the model will see, hence it's a better measure of the model's performance when launched in production (aka, why we use it as the final metrics), and its use in refitting the model will create a model which performs better on the production traffic.